### PR TITLE
previous_dtype is now inferred from F.linear's result output type.

### PR DESCRIPTION
**Pull request by MFajcik**
Fixes the issue from https://github.com/huggingface/peft/issues/971

### CODE REVIEW

**Comment by MFajcik** [comment]
Done. Also, I tested it, successfully running my experiment with GPT-2 LORA fine-tuning, with original model computation running in bf16 as intended.

**Review by BenjaminBossan — COMMENTED** [review]
Thanks for providing this fix and for testing it. I ran the GPU tests locally and they also passed (but I can't test with bf16). Do you happen to have a quick test at the ready to show that it works with bf16 and that doesn't involve inspecting intermediate results (as in the original issue)?

Regarding the PR, could you please remove the empty line 243, it's not required anymore. Also, should the `forward` methods of `Embedding` and `Conv2d` also be changed accordingly?

**Comment by BenjaminBossan** [comment]
@MFajcik Are you still working on this?

**Comment by BenjaminBossan** [comment]
@MFajcik Friendly reminder about this PR :)

There are some merge conflicts now due to a recent change. Let me know if you need help resolving it.

**Comment by MFajcik** [comment]
@BenjaminBossan Yeah I didn't forget about this. I just don't have time right now. 
I will try to resolve these next week.

**Comment by BenjaminBossan** [comment]
Fantastic, thanks. I don't want to put any pressure on you, take the time you need, I just wanted to ping you because sometimes people just forget :)

**Comment by MFajcik** [comment]
@BenjaminBossan I merged changes, and reapplied fix for Conv2D, Linear and Embeddings. I also ran make format, and added 1 test covering the fix for LoRA.  Is that ok?

**Comment by BenjaminBossan** [comment]
@MFajcik Thanks for making the updates. The quality check still fails, could you please re-run `make style`? If that doesn't report back anything, maybe some package versions are outdated and need upgrading.

> added 1 test covering the fix for LoRA

Thanks for adding the test. Probably we can move it to some of the existing files, but I'll take a closer look later and make a suggestion.

**Comment by MFajcik** [comment]
> @MFajcik Thanks for making the updates. The quality check still fails, could you please re-run `make style`? If that doesn't report back anything, maybe some package versions are outdated and need upgrading.
> 
> > added 1 test covering the fix for LoRA
> 
> Thanks for adding the test. Probably we can move it to some of the existing files, but I'll take a closer look later and make a suggestion.

Yeah, I forgot to run make style after making the test again... Should be okay now.

**Comment by BenjaminBossan** [comment]
@MFajcik Code quality checks are still failing:

> tests/test_autocast_torchcompatibility_lora.py:141:20: F841 Local variable `labels` is assigned to but never used

**Comment by BenjaminBossan** [comment]
Friendly ping @MFajcik

**Comment by MFajcik** [comment]
now? @BenjaminBossan

File: tests/test_autocast_torchcompatibility_lora.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+        norm_output = norm_output.unsqueeze(1)
+        conv_output = self.conv2d_transform(norm_output)
+
+        # Remove batch size dimension
+        conv_output = conv_output.squeeze(1)
+
+        return conv_output
+
+
+class TestAutoCast(unittest.TestCase):
```
Comment (younesbelkada) [inline tests/test_autocast_torchcompatibility_lora.py:120]: This test seems to require a GPU, can you add the `require_torch_gpu` decorator here? https://github.com/huggingface/peft/blob/3708793ba9a659a92fd9c20a0d8fe1bb94e0c5aa/tests/testing_utils.py#L25

**Review by younesbelkada — COMMENTED** [review]
Thanks, I left a single comment with respect to failing tests

File: tests/test_autocast_torchcompatibility_lora.py (hunk 1 above)
Comment (MFajcik) [reply tests/test_autocast_torchcompatibility_lora.py:123]: done.

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
Thanks a lot @MFajcik for making LoRA work correctly with AMP. This looks pretty good already. I have a couple of change requests and suggestions for improvement, please check them out.

In terms of the general approach, now that we use

```python
result = self.base_layer(x, *args, **kwargs)
```

I think it's much safer to make this change than at the time the issue was originally discussed. This way, we should have a guarantee that LoRA does not change the dtype of the output.

**Comment by BenjaminBossan** [comment]
@MFajcik Do you still plan to work on this? :)

**Comment by MFajcik** [comment]
> @MFajcik Do you still plan to work on this? :)

Hopefully this month sometimes :)

**Comment by MFajcik** [comment]
@BenjaminBossan I think I adressed all your comments now.

File: tests/test_gpu_examples.py [hunk 2]
```diff
@@ -1378,3 +1378,110 @@ def test_model_loaded_in_float16_working(self):
...
+        conv_output = self.conv2d_transform(norm_output)
+
+        # Remove batch size dimension
+        conv_output = conv_output.squeeze(1)
+
+        return conv_output
+
+
+@require_torch_gpu
+class TestAutoCast(unittest.TestCase):
```
Comment (BenjaminBossan) [inline tests/test_gpu_examples.py:1432]: Could you please add a short comment about what is being tested here?

**Review by BenjaminBossan — APPROVED** [review]
Thanks a lot @MFajcik for providing the fix and adding tests for it. From my point of view, this looks good, but I'd like to have another review by @younesbelkada or @pacman100 to be 100% sure.

I saw that other methods probably need to be changed in the same way as done in this PR (LoHa, LoKr, poly, oft, IA³, `lora.tp_layer.LoraParallelLinear`), but that can be done in a subsequent PR.

File: tests/test_gpu_examples.py (hunk 2 above)
Comment (MFajcik) [reply tests/test_gpu_examples.py:1432]: Here.

**Review by pacman100 — APPROVED** [review]
Thank you @MFajcik for the fixes and @BenjaminBossan for all the discussions and suggestions!

**Comment by BenjaminBossan** [comment]
Could you please run `make style` too :)

**Pull request merged**